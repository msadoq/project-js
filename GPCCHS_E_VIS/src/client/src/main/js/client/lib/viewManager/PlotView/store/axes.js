// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #5828 : 16/03/2017 : Add PlotView and DynamicView configurationReducer in
//  viewManager
// VERSION : 1.1.2 : DM : #5828 : 04/04/2017 : Remove old configuration reducers from
//  reducers/views/configuration
// VERSION : 1.1.2 : DM : #5828 : 26/04/2017 : Ported 1.1.0 patch to dev branch. EP Drag & drop
//  auto-axis-creation.
// VERSION : 1.1.2 : DM : #5828 : 27/04/2017 : Fixed tests - auto axis creation when new
//  EntryPoint.
// VERSION : 1.1.2 : DM : #6829 : 27/06/2017 : Plot axes log settings stored in store and
//  documents.
// VERSION : 2.0.0 : FA : #8082 : 11/12/2017 : Plot view refresh problem // Lorsqu'on cree un entry
//  point...
// END-HISTORY
// ====================================================================

import __ from 'lodash/fp';

/*
 * All of this functions are used by the configuration reducer
*/

export function updateAxis(stateConf, action) {
  const { axis, axisId } = action.payload;
  if (!axis || !axisId) {
    return stateConf;
  }
  return __.set(`axes[${axisId}]`, { ...axis, id: axisId }, stateConf);
}

export function addAxis(stateConf, action) {
  const { axis } = action.payload;
  if (!axis || !axis.label) {
    return stateConf;
  }
  const axisId = axis.id || getUniqueAxisId(stateConf, axis.label);
  return __.set(`axes[${axisId}]`, { ...axis, id: axisId }, stateConf);
}

export function removeAxis(stateConf, action) {
  const { axisId } = action.payload;
  if (!axisId) {
    return stateConf;
  }
  return __.update('axes', __.omit(axisId), stateConf);
}

function getUniqueAxisId(stateConf, label) {
  const id = __.snakeCase(__.trim(__.deburr(label)));
  // check id isn't already defined
  let isUnique = false;
  let index = 1;
  let finalId = id;
  const { axes } = stateConf;
  while (!isUnique) {
    if (!axes[finalId]) {
      isUnique = true;
    } else {
      finalId = id.concat('_', index);
      index += 1;
    }
  }
  return finalId;
}

export function createAxis(stateConf, label, unit) {
  return {
    autoLimits: true,
    showTicks: true,
    autoTick: true,
    showAxis: true,
    showLabels: true,
    max: 300,
    min: -300,
    logarithmic: false,
    logSettings: {
      min: 0.1,
      max: 1000000000,
      base: 10,
    },
    label,
    unit,
    id: getUniqueAxisId(stateConf, label),
  };
}

// This is not a reducer because it returns pair of axes
export const getYAxis = (stateConf, action) => {
  const { entryPoint } = action.payload;
  const { connectedData } = entryPoint;
  const { axes } = stateConf;
  const axisYKey = Object.keys(axes).find(k => axes[k].id && axes[k].id === connectedData.axisId);
  const axisY = axisYKey ? axes[axisYKey] : null;
  const finalY = axisY || createAxis(stateConf, entryPoint.name, connectedData.unit);
  const prefixY = finalY.id === 'time' ? 'Y:' : '';

  return { ...finalY, id: `${prefixY}${finalY.id}` };
};
