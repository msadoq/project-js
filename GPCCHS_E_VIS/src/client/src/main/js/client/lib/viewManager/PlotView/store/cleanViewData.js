/* eslint-disable no-continue */
import _difference from 'lodash/difference';
import _get from 'lodash/get';
import _isEmpty from 'lodash/isEmpty';
import _head from 'lodash/head';
import _last from 'lodash/last';
import _isEqual from 'lodash/isEqual';
import _findIndex from 'lodash/findIndex';
import _findLastIndex from 'lodash/findLastIndex';
import _omit from 'lodash/omit';

/** ************************************************
 * Clean viewData for current viewData
 * @param current view data State
 * @param old current view definition
 * @param new current view definition
 * @param old expected intervals for all entry points
 * @param new expected intervals for all entry points
 * @return cleaned state for current view
/** ************************************************/
export default function cleanCurrentViewData(
  currentState,
  oldViewFromMap,
  newViewFromMap,
  oldIntervals,
  newIntervals) {
  // Check if viewMap has changed
  if (_isEqual(newViewFromMap, oldViewFromMap) && _isEqual(oldIntervals, newIntervals)) {
    return currentState;
  }
  // new visible view
  if (!oldViewFromMap || !currentState || _isEmpty(currentState.indexes)) {
    return currentState;
  }
  let newState = currentState;
  // invisible view
  if (!newViewFromMap) {
    return {
      indexes: {},
      lines: {},
      min: {},
      max: {},
      minTime: {},
      maxTime: {},
    };
  }
  // entry point updates
  const oldEntryPoints = oldViewFromMap.entryPoints;
  const newEntryPoints = newViewFromMap.entryPoints;
  const epNames = Object.keys(oldEntryPoints);
  epNames.forEach((epName) => {
    const oldEp = oldEntryPoints[epName];
    const newEp = newEntryPoints[epName];
    // check if only label has changed
    if (!newEp) {
      const diff = _difference(Object.keys(newEntryPoints), Object.keys(oldEntryPoints));
      let newLabel;
      if (diff.length) {
        diff.forEach((name) => {
          if (newEntryPoints[name].id === oldEntryPoints[name].id) {
            newLabel = name;
          }
        });
      }
      if (newLabel) {
        newState = updateEpLabel(newState, epName, newLabel);
        return;
      }
    }
    // removed entry point if invalid
    // EP definition modified: remove entry point from viewData
    if (!newEp || (newEp.error && newEp.error !== oldEp.error)
      || oldEp.fieldX !== newEp.fieldX || oldEp.fieldY !== newEp.fieldY
      || oldEp.remoteId !== newEp.remoteId) {
      // newState = u({ indexes: u.omit(epName), lines: u.omit(epName) }, newState);
      newState = { ...newState,
        indexes: _omit(newState.indexes, epName),
        lines: _omit(newState.lines, epName),
        min: _omit(newState.min, epName),
        minTime: _omit(newState.minTime, epName),
        max: _omit(newState.max, epName),
        maxTime: _omit(newState.maxTime, epName),
      };
      return;
    }
    // Case of point already in error
    if (newEp.error) {
      return;
    }
    // update on expected interval
    const oldInterval = _get(oldIntervals, [oldEp.remoteId, oldEp.localId, 'expectedInterval']);
    const newInterval = _get(newIntervals, [oldEp.remoteId, oldEp.localId, 'expectedInterval']);
    if (!oldInterval || !newInterval ||
      oldInterval[0] !== newInterval[0] || oldInterval[1] !== newInterval[1]) {
      const lower = newInterval[0] + newEp.offset;
      const upper = newInterval[1] + newEp.offset;
      const newData = removeViewDataByEp(newState, epName, lower, upper);
      // Update of min and max if needed
      newState = scanForMinAndMax(newData);
    }
  });
  return newState;
}

export function updateEpLabel(viewData, oldLabel, newLabel) {
  if (!oldLabel || !newLabel || oldLabel === newLabel) {
    return viewData;
  }
  // unknown label
  if (!viewData.lines[oldLabel]) {
    return viewData;
  }

  const newState = { ...viewData,
    lines: _omit(viewData.lines, oldLabel),
    indexes: _omit(viewData.indexes, oldLabel),
    min: _omit(viewData.min, oldLabel),
    minTime: _omit(viewData.minTime, oldLabel),
    max: _omit(viewData.max, oldLabel),
    maxTime: _omit(viewData.maxTime, oldLabel),
  };

  return { ...newState,
    lines: { ...newState.lines, [newLabel]: viewData.lines[oldLabel] },
    indexes: { ...newState.indexes, [newLabel]: viewData.indexes[oldLabel] },
    min: { ...newState.min, [newLabel]: viewData.min[oldLabel] },
    max: { ...newState.max, [newLabel]: viewData.max[oldLabel] },
    minTime: { ...newState.minTime, [newLabel]: viewData.minTime[oldLabel] },
    maxTime: { ...newState.maxTime, [newLabel]: viewData.maxTime[oldLabel] },
  };
}

// function used to update min and max when no data is added to state
export function scanForMinAndMax(viewDataState) {
  if (!viewDataState || !viewDataState.indexes || !Object.keys(viewDataState.indexes).length) {
    return viewDataState;
  }
  const minVal = {};
  const maxVal = {};
  const minTimeVal = {};
  const maxTimeVal = {};
  // loop on EP names
  const epNames = Object.keys(viewDataState.indexes);
  epNames.forEach((epName) => {
    // get interval to evaluate the min and max validity
    const lower = _head(viewDataState.indexes[epName]);
    const upper = _last(viewDataState.indexes[epName]);
    // Check validity of current extrema
    if (viewDataState.minTime[epName] >= lower && viewDataState.minTime[epName] <= upper
      && viewDataState.maxTime[epName] >= lower && viewDataState.maxTime[epName] <= upper) {
      // nothing to update
      return;
    }

    // Scan state values
    viewDataState.lines[epName].forEach((point) => {
      if (!point.value) {
        return;
      }
      if (!minVal[epName] || minVal[epName] >= point.value) {
        minVal[epName] = point.value;
        minTimeVal[epName] = point.masterTime;
      }
      if (!maxVal[epName] || maxVal[epName] <= point.value) {
        maxVal[epName] = point.value;
        maxTimeVal[epName] = point.masterTime;
      }
    });
  });
  // Nothing to update
  if (_isEmpty(minVal) && _isEmpty(maxVal)) {
    return viewDataState;
  }
  return { ...viewDataState,
    min: Object.assign({}, viewDataState.min, minVal),
    max: Object.assign({}, viewDataState.max, maxVal),
    minTime: Object.assign({}, viewDataState.minTime, minTimeVal),
    maxTime: Object.assign({}, viewDataState.maxTime, maxTimeVal),
  };
}

export function removeViewDataByEp(viewData, epName, lower, upper) {
  if (lower > upper) {
    // unpredictable usage
    return { indexes: {}, lines: {}, min: {}, max: {}, minTime: {}, maxTime: {} };
  }
  // keep everything
  if (!viewData || !viewData.indexes || !viewData.indexes[epName]) {
    // state contains no data
    return viewData;
  }
  if (viewData.indexes[epName][0] >= lower && _last(viewData.indexes[epName]) <= upper) {
    // all points of entryPoint are in visuWindow
    return viewData;
  }

  // drop everything: keep min and max for PlotView drawing
  if (viewData.indexes[epName][0] > upper || _last(viewData.indexes[epName]) < lower) {
    return { ...viewData,
      indexes: _omit(viewData.indexes, epName),
      lines: _omit(viewData.lines, epName),
      min: _omit(viewData.min, epName),
      minTime: _omit(viewData.minTime, epName),
      max: _omit(viewData.max, epName),
      maxTime: _omit(viewData.maxTime, epName),
    };
  }
  // cut
  const iLower = _findIndex(viewData.indexes[epName], t => t >= lower);
  let iUpper = _findLastIndex(viewData.indexes[epName], t => t <= upper);
  iUpper = (iUpper === -1) ? viewData.indexes[epName].length - 1 : iUpper;
  return { ...viewData,
    indexes: {
      ...viewData.indexes,
      [epName]: viewData.indexes[epName].slice(iLower, iUpper + 1),
    },
    lines: {
      ...viewData.lines,
      [epName]: viewData.lines[epName].slice(iLower, iUpper + 1),
    },
  };
}
