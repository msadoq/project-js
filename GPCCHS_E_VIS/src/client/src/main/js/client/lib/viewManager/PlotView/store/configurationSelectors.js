import _ from 'lodash/fp';
import { createSelector } from 'reselect';

import { getViewConfiguration } from '../../../store/reducers/views';

export const getAxes = createSelector(
  getViewConfiguration,
  _.get('axes')
);

export const getShowYAxes = createSelector(
  getViewConfiguration,
  _.get('showYAxes')
);
