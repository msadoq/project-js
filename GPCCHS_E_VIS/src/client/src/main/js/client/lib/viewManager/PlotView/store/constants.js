import _ from 'lodash/fp';

export function addConstant(stateConf, action) {
  const { constant } = action.payload;
  if (!constant || !constant.label) {
    return stateConf;
  }
  const constantId = constant.id || getUniqueConstantId(stateConf, constant.label);
  return _.set(`constants[${constantId}]`, { ...constant, id: constantId }, stateConf);
}

export function updateConstant(stateConf, action) {
  const { constant, constantId } = action.payload;
  if (!constant || !constantId) {
    return stateConf;
  }
  return _.set(`constants[${constantId}]`, { ...constant, id: constantId }, stateConf);
}

export function removeConstant(stateConf, action) {
  const { constantId } = action.payload;
  if (!constantId) {
    return stateConf;
  }
  return _.update('constants', constants => _.omit(constantId, constants), stateConf);
}

function getUniqueConstantId(stateConf, label) {
  const id = _.snakeCase(_.trim(_.deburr(label)));
  // check id isn't already defined
  let isUnique = false;
  let index = 1;
  let finalId = id;
  const { constants } = stateConf;
  while (!isUnique) {
    if (!constants[finalId]) {
      isUnique = true;
    } else {
      finalId = id.concat('_', index);
      index += 1;
    }
  }
  return finalId;
}
