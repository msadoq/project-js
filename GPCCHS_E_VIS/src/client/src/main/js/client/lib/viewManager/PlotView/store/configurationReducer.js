import _ from 'lodash/fp';

import { getYAxis, updateAxis, addAxis, removeAxis } from './axes';
import * as types from '../../../store/types';

const removeElementIn = (key, index, state) => _.update(key, _.pullAt(index), state);

const addElementIn = (key, val, state) => _.update(key, x => _.compact(_.concat(x, val)), state);

/* eslint-disable complexity, "DV6 TBC_CNES Redux reducers should be implemented as switch case" */
export default (stateConf = { search: '', showLegend: true }, action) => {
  switch (action.type) {
    case types.WS_VIEW_UPDATE_LEGEND:
      return _.set('legend', action.payload.legend, stateConf);
    case types.WS_VIEW_UPDATE_SHOWYAXES:
      return _.set('showYAxes', action.payload.showYAxes, stateConf);
    case types.WS_VIEW_UPDATE_GRID:
      return _.set(`grids[${action.payload.index}]`, action.payload.grid, stateConf);
    case types.WS_VIEW_TOGGLE_LEGEND:
      return _.set('showLegend', action.payload.flag, stateConf);
    case types.WS_VIEW_ADD_GRID:
      return addElementIn('grids', action.payload.grid, stateConf);
    case types.WS_VIEW_REMOVE_GRID:
      return removeElementIn('grids', action.payload.index, stateConf);
    case types.WS_VIEW_UPDATE_MARKER:
      return _.set(`markers[${action.payload.index}]`, action.payload.marker, stateConf);
    case types.WS_VIEW_ADD_MARKER:
      return addElementIn('markers', action.payload.marker, stateConf);
    case types.WS_VIEW_REMOVE_MARKER:
      return removeElementIn('markers', action.payload.index, stateConf);
    case types.WS_VIEW_ADD_ENTRYPOINT: {
      const axisY = getYAxis(stateConf, action);
      return {
        ...stateConf,
        entryPoints: [
          ...stateConf.entryPoints,
          {
            ...action.payload.entryPoint,
            connectedData: {
              ...(action.payload.entryPoint.connectedData),
              axisId: axisY.id,
            },
          },
        ],
        axes: {
          ...stateConf.axes,
          [axisY.id]: axisY,
        },
      };
    }
    case types.WS_VIEW_REMOVE_ENTRYPOINT: {
      const entryPoints = stateConf.entryPoints;
      const getAllConnectedAxisIds = eps => _.pluck('connectedData.axisId', eps);
      const refreshAxes = _.pick(getAllConnectedAxisIds(entryPoints));
      return _.update('axes', refreshAxes, stateConf);
    }
    case types.WS_VIEW_UPDATE_AXIS: {
      return updateAxis(stateConf, action);
    }
    case types.WS_VIEW_ADD_AXIS:
      return addAxis(stateConf, action);
    case types.WS_VIEW_REMOVE_AXIS:
      return removeAxis(stateConf, action);
    case types.WS_VIEW_UPDATE_EDITOR_SEARCH:
      if (action.payload.search !== stateConf.search) {
        return _.set('search', action.payload.search, stateConf);
      }
      return stateConf;
    default:
      return stateConf;
  }
};
