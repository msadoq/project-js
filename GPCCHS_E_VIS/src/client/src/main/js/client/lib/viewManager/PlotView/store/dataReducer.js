// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #5828 : 20/03/2017 : Creation of data store for plotView
// VERSION : 1.1.2 : DM : #5828 : 24/03/2017 : Fix crash vima when load a blank workspace
// VERSION : 1.1.2 : DM : #5828 : 03/04/2017 : Add some eslint relaxation rules
// VERSION : 1.1.2 : DM : #6302 : 03/04/2017 : Add comment and fix coding convetions warning and un-needed relaxations
// VERSION : 1.1.2 : DM : #5828 : 05/04/2017 : update viewData when closing a page
// VERSION : 1.1.2 : DM : #5828 : 11/04/2017 : Implement automation for data reducers in viewManager
// VERSION : 1.1.2 : DM : #6700 : 06/07/2017 : Rename documentManager actions . .
// VERSION : 1.1.2 : DM : #6700 : 01/08/2017 : Branch full cycle mechanism for rangeData
// VERSION : 1.1.2 : DM : #6700 : 02/08/2017 : Update unit tests for Plot View store
// VERSION : 1.1.2 : DM : #6700 : 03/08/2017 : split of viewData cleaning in dataReducer for plot
// VERSION : 1.1.2 : DM : #6700 : 03/08/2017 : Merge branch 'dev' into dbrugne-data
// END-HISTORY
// ====================================================================

import _omit from 'lodash/omit';

import * as types from 'store/types';
import * as constants from 'viewManager/constants';

import cleanCurrentViewData from './cleanViewData';
import { viewRangeAdd, selectDataPerView } from './viewDataUpdate';

const initialState = {
  indexes: {},
  lines: {},
  min: {},
  max: {},
  minTime: {},
  maxTime: {},
};

/* eslint-disable complexity, "DV6 TBC_CNES Redux reducers should be implemented as switch case" */
export default function plotViewData(state = {}, action) {
  switch (action.type) {
    case types.DATA_REMOVE_ALL_VIEWDATA:
    case types.HSC_CLOSE_WORKSPACE:
      return {};
    case types.WS_VIEW_OPENED:
    case types.WS_VIEW_ADD_BLANK:
      if (action.payload.view.type !== constants.VM_VIEW_PLOT) {
        return state;
      }
      return { ...state, [action.payload.view.uuid]: initialState };
    case types.WS_PAGE_OPENED:
    case types.WS_WORKSPACE_OPENED:
      {
        const { views } = action.payload;
        if (!views) {
          return state;
        }
        const newState = {};
        views.forEach((view) => {
          if (view.type !== constants.VM_VIEW_PLOT) {
            return;
          }
          newState[view.uuid] = initialState;
        });
        return { ...state, ...newState };
      }
    case types.WS_VIEW_CLOSE: {
      const { viewId } = action.payload;
      if (state[viewId]) {
        return _omit(state, viewId);
      }
      return state;
    }
    case types.WS_PAGE_CLOSE: {
      const { viewIds } = action.payload;
      if (!viewIds.length) {
        return state;
      }
      let newState = state;
      viewIds.forEach((viewId) => {
        if (state[viewId]) {
          newState = _omit(newState, viewId);
        }
      });
      return newState;
    }
    case types.INJECT_DATA_RANGE: {
      const { dataToInject, newViewMap,
              newExpectedRangeIntervals, configurations } = action.payload;
      const { PlotViewConfiguration } = configurations;
      const dataKeys = Object.keys(dataToInject);
      // If nothing changed and no data to import, return state
      if (!dataKeys.length) {
        return state;
      }
      // since now, state will changed
      let newState = state;
      const viewIds = Object.keys(state);
      for (let i = 0; i < viewIds.length; i += 1) {
        const viewId = viewIds[i];
        // Data Selection
        const epSubState =
          selectDataPerView(newViewMap[viewId],
                            PlotViewConfiguration[viewId],
                            newExpectedRangeIntervals,
                            dataToInject,
                            state[viewId]);
        if (Object.keys(epSubState).length !== 0) {
          // Data injection
          const viewState = viewRangeAdd(newState[viewId], epSubState);
          if (viewState !== newState[viewId]) {
            newState = { ...newState, [viewId]: viewState };
          }
        }
      }
      return newState || {};
    }
    case types.WS_VIEWDATA_CLEAN: {
      const { previousDataMap, dataMap, configuration } = action.payload;
      const { PlotViewConfiguration } = configuration;
      // since now, state will changed
      let newState = state;
      const viewIds = Object.keys(state);
      for (let i = 0; i < viewIds.length; i += 1) {
        const viewId = viewIds[i];
        const viewData = state[viewId];
        // Cleaning
        const subState = cleanCurrentViewData(
          viewData,
          previousDataMap.perView[viewId],
          dataMap.perView[viewId],
          previousDataMap.expectedRangeIntervals,
          dataMap.expectedRangeIntervals,
          PlotViewConfiguration[viewId]);
        if (subState !== viewData) {
          newState = { ...newState, [viewId]: subState };
        }
      }
      return newState;
    }
    default:
      return state;
  }
}

export const getPlotViewData = state => state.PlotViewData;

export const getData = (state, { viewId }) => state.PlotViewData[viewId];
