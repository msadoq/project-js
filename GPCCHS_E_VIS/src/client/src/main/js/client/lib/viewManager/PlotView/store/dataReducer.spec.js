// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #5828 : 20/03/2017 : Creation of data store for plotView
// VERSION : 1.1.2 : DM : #3622 : 22/03/2017 : Update viewData organization for last structure + cleaning
// VERSION : 1.1.2 : DM : #5828 : 05/04/2017 : update viewData when closing a page
// VERSION : 1.1.2 : FA : #6670 : 12/06/2017 : Apply jest-codemods for chai-should + repair lots of tests
// VERSION : 1.1.2 : DM : #5828 : 14/06/2017 : Refactor Jest test to replace it() with test() calls
// VERSION : 1.1.2 : DM : #5828 : 14/06/2017 : Move common/log and common/parameters in client/
// VERSION : 1.1.2 : FA : #6670 : 16/06/2017 : Move and rename jest.js in jest/setup.js + test.js in jest/index.js
// VERSION : 1.1.2 : DM : #6700 : 06/07/2017 : Rename documentManager actions . .
// VERSION : 1.1.2 : DM : #6700 : 02/08/2017 : Update unit tests for Plot View store
// END-HISTORY
// ====================================================================

import { freezeMe } from 'common/jest';
import * as types from 'store/types';
import state from 'common/jest/stateTest';
import dataMapGenerator from 'dataManager/map';
import plotViewData from './dataReducer';

describe('viewManager/PlotView/store/dataReducer', () => {
  test('DATA_REMOVE_ALL_VIEWDATA', () => {
    const action = { type: types.DATA_REMOVE_ALL_VIEWDATA };
    expect(plotViewData(state.PlotViewData, action))
      .toEqual({});
  });
  test('HSC_CLOSE_WORKSPACE', () => {
    expect(plotViewData(state.PlotViewData, { type: types.HSC_CLOSE_WORKSPACE }))
      .toEqual({});
  });
  test('WS_VIEW_OPENED', () => {
    const action = {
      type: types.WS_VIEW_OPENED,
      payload: { view: { type: 'TextView', uuid: 'myText' } },
    };
    expect(plotViewData(freezeMe({}), action))
      .toEqual({});
    action.payload.view = { type: 'PlotView', uuid: 'myPlot' };
    expect(plotViewData(freezeMe({}), action))
      .toEqual({
        myPlot: {
          indexes: {}, lines: {}, min: {}, max: {}, minTime: {}, maxTime: {},
        },
      });
  });
  test('WS_VIEW_ADD_BLANK', () => {
    const action = {
      type: types.WS_VIEW_ADD_BLANK,
      payload: { view: { type: 'TextView', uuid: 'myText' } },
    };
    expect(plotViewData(freezeMe({}), action))
      .toEqual({});
    action.payload.view = { type: 'PlotView', uuid: 'myPlot' };
    expect(plotViewData(freezeMe({}), action))
      .toEqual({
        myPlot: {
          indexes: {}, lines: {}, min: {}, max: {}, minTime: {}, maxTime: {},
        },
      });
  });
  test('WS_PAGE_OPENED', () => {
    const action = {
      type: types.WS_PAGE_OPENED,
      payload: {
        views:
          [{ type: 'PlotView', uuid: 'myPlot' }, { type: 'TextView', uuid: 'myText' }],
      },
    };
    expect(plotViewData(freezeMe({}), action))
      .toEqual({
        myPlot: {
          indexes: {}, lines: {}, min: {}, max: {}, minTime: {}, maxTime: {},
        },
      });
  });
  test('WS_WORKSPACE_OPENED', () => {
    const action = {
      type: types.WS_WORKSPACE_OPENED,
      payload: {
        views:
          [{ type: 'PlotView', uuid: 'myPlot' }, { type: 'TextView', uuid: 'myText' }],
      },
    };
    expect(plotViewData(freezeMe({}), action))
      .toEqual({
        myPlot: {
          indexes: {}, lines: {}, min: {}, max: {}, minTime: {}, maxTime: {},
        },
      });
  });
  test('WS_VIEW_CLOSE', () => {
    const action = { type: types.WS_VIEW_CLOSE, payload: { viewId: 'myText' } };
    expect(plotViewData(freezeMe({}), action))
      .toEqual({});
    const frozen = freezeMe({
      myPlot: {
        indexes: {}, lines: {}, min: {}, max: {}, minTime: {}, maxTime: {},
      },
    });
    expect(plotViewData(freezeMe(frozen), action))
      .toBe(frozen);
    action.payload.viewId = 'myPlot';
    expect(plotViewData(frozen, action))
      .toEqual({});
  });
  test('WS_PAGE_CLOSE', () => {
    const action = { type: types.WS_PAGE_CLOSE, payload: { viewIds: ['myPlot', 'myText'] } };
    expect(plotViewData(freezeMe({ myPlot: {}, myOtherPlot: {} }), action))
      .toEqual({ myOtherPlot: {} });
  });
  test('Unknown action', () => {
    const action = { type: types.UNKNOWN, payload: { viewId: 'myPlot' } };
    const frozen = freezeMe({
      myPlot: {
        indexes: {}, lines: {}, min: {}, max: {}, minTime: {}, maxTime: {},
      },
    });
    expect(plotViewData(freezeMe(frozen), action))
      .toBe(frozen);
  });
  describe('INJECT_DATA_RANGE', () => {
    const dataMap = dataMapGenerator(state);
    const payload = {
      'Reporting.TMMGT_BC_VIRTCHAN3<ReportingParameter>:0:4:extractedValue.<.100': {},
      'Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1': {},
      tbdId3: {},
    };
    for (let j = 100120; j < 100920; j += 100) {
      payload['Reporting.TMMGT_BC_VIRTCHAN3<ReportingParameter>:0:4:extractedValue.<.100'][j] = {
        extractedValue: { type: 'uinteger', value: (j / 10000) + 1 },
        rawValue: { type: 'uinteger', value: j + 2 },
        convertedValue: { type: 'uinteger', value: j + 3 },
        referenceTimestamp: { type: 'time', value: j },
        time: { type: 'time', value: j + 0.2 },
        groundDate: { type: 'time', value: j + 0.2 },
      };

      payload['Reporting.ATT_BC_REVTCOUNT1<ReportingParameter>:0:1'][j] =
        payload['Reporting.TMMGT_BC_VIRTCHAN3<ReportingParameter>:0:4:extractedValue.<.100'][j];
      payload.tbdId3[j] =
        payload['Reporting.TMMGT_BC_VIRTCHAN3<ReportingParameter>:0:4:extractedValue.<.100'][j];
    }

    test('ok', () => {
      const frozen = freezeMe(state.PlotViewData);
      const action = {
        type: types.INJECT_DATA_RANGE,
        payload: {
          oldViewMap: dataMap.perView,
          newViewMap: dataMap.perView,
          oldExpectedRangeIntervals: dataMap.expectedRangeIntervals,
          newExpectedRangeIntervals: dataMap.expectedRangeIntervals,
          dataToInject: payload,
          configurations: { PlotViewConfiguration: state.PlotViewConfiguration },
        },
      };

      const newState = plotViewData(frozen, action);
      expect(newState)
        .toMatchSnapshot();
    });
    test('nothing to add', () => {
      const frozen = freezeMe(state.PlotViewData);
      const action = {
        type: types.INJECT_DATA_RANGE,
        payload: {
          oldViewMap: dataMap.perView,
          newViewMap: dataMap.perView,
          oldExpectedRangeIntervals: dataMap.expectedRangeIntervals,
          newExpectedRangeIntervals: dataMap.expectedRangeIntervals,
          dataToInject: {},
          configurations: { PlotViewConfiguration: state.PlotViewConfiguration },
        },
      };
      const newState = plotViewData(frozen, action);
      expect(newState)
        .toBe(frozen);
    });
  });
  describe('INJECT_DATA_OBSOLETE_EVENT', () => {
    const dataMap = dataMapGenerator(state);
    test('nothing to add', () => {
      const frozen = freezeMe(state.PlotViewData);
      const action = {
        type: types.INJECT_DATA_OBSOLETE_EVENT,
        payload: {
          newViewMap: dataMap.perView,
          dataToInject: {},
        },
      };
      const newState = plotViewData(frozen, action);
      expect(newState)
        .toBe(frozen);
    });
    test('add an obsolete event for view plot1 and entrypoint ATT_BC_REVTCOUNT1 ', () => {
      const frozen = freezeMe(state.PlotViewData);
      const action = {
        type: types.INJECT_DATA_OBSOLETE_EVENT,
        payload: {
          newViewMap: dataMap.perView,
          dataToInject: {
            'ATT_BC_REVTCOUNT1:0:1:::': {
              10: {
                eventDate: 10,
              },
            },
          },
        },
      };
      const newState = plotViewData(frozen, action);
      expect(newState.plot1.obsoleteEvents)
        .toEqual({
          'ATT_BC_REVTCOUNT1:0:1:::': {
            10: {
              eventDate: 10,
            },
          },
        });
    });
  });
});
