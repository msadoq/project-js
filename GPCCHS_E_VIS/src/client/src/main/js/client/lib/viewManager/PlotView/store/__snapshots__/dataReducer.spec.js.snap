// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewManager/PlotView/store/dataReducer INJECT_DATA_RANGE ok 1`] = `
Object {
  "plot1": Object {
    "indexes": Object {
      "ATT_BC_REVTCOUNT1": Array [
        100020,
        200020,
        300020,
        400020,
        500020,
      ],
      "TMMGT_BC_VIRTCHAN3": Array [
        100020,
        200020,
        300020,
        400020,
        500020,
      ],
    },
    "lines": Object {
      "ATT_BC_REVTCOUNT1": Object {
        "100020": Object {
          "color": "darkred",
          "masterTime": 100020,
          "symbol": 149.30028143545633,
          "value": 149.30028143545633,
          "x": 100000,
        },
        "200020": Object {
          "color": "blue",
          "masterTime": 200020,
          "symbol": 154.3002814355,
          "value": 154.3002814355,
          "x": 200000,
        },
        "300020": Object {
          "color": "green",
          "masterTime": 300020,
          "symbol": 159.3002814355,
          "value": 159.3002814355,
          "x": 300000,
        },
        "400020": Object {
          "color": "green",
          "masterTime": 400020,
          "symbol": 164.3002814355,
          "value": 164.3002814355,
          "x": 400000,
        },
        "500020": Object {
          "color": "green",
          "masterTime": 500020,
          "symbol": 169.3002814355,
          "value": 169.3002814355,
          "x": 500000,
        },
      },
      "TMMGT_BC_VIRTCHAN3": Object {
        "100020": Object {
          "color": "darkred",
          "masterTime": 100020,
          "symbol": 139.30028143545633,
          "value": 139.30028143545633,
          "x": 100000,
        },
        "200020": Object {
          "color": "blue",
          "masterTime": 200020,
          "symbol": 144.3002814355,
          "value": 144.3002814355,
          "x": 200000,
        },
        "300020": Object {
          "color": "green",
          "masterTime": 300020,
          "symbol": 149.3002814355,
          "value": 149.3002814355,
          "x": 300000,
        },
        "400020": Object {
          "color": "green",
          "masterTime": 400020,
          "symbol": 154.3002814355,
          "value": 154.3002814355,
          "x": 400000,
        },
        "500020": Object {
          "color": "green",
          "masterTime": 500020,
          "symbol": 159.3002814355,
          "value": 159.3002814355,
          "x": 500000,
        },
      },
    },
    "max": Object {
      "ATT_BC_REVTCOUNT1": 169.3002814355,
      "TMMGT_BC_VIRTCHAN3": 159.30028143545633,
    },
    "maxTime": Object {
      "ATT_BC_REVTCOUNT1": 500020,
      "TMMGT_BC_VIRTCHAN3": 500020,
    },
    "min": Object {
      "ATT_BC_REVTCOUNT1": 149.30028143545633,
      "TMMGT_BC_VIRTCHAN3": 139.30028143545633,
    },
    "minTime": Object {
      "ATT_BC_REVTCOUNT1": 100020,
      "TMMGT_BC_VIRTCHAN3": 100020,
    },
  },
}
`;
